### PR TITLE
Initial docs for sync local changes with unison

### PR DESCRIPTION
**Pull request by markbiek**
## Proposed changes:
Based on the Jetpack workflow discussion in pdWQjU-ur-p2, this adds some initial documentation for using Unison file-sync to sync local Jetpack changes to a remote server (like a sandbox) for a more real-time workflow.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Testing instructions:

NA

### CODE REVIEW

File: docs/development-environment.md [hunk 1]
```diff
@@ -192,6 +190,50 @@ There are different types of builds:
...
 
+* ### Syncing local changes with Unison
+  
+  In some cases, you may need to test Jetpack (jetpack-mu-wpcom, in particular) changes by syncing your changes to another machine (rather than using Docker). This outlines a strategy for syncing changes in real-time using the [Unison](https://github.com/bcpierce00/unison) file sync tool combines with the [unison-fsmonitor](https://github.com/benesch/unison-fsmonitor).
+
+  This approach may be especially useful for Automatticians who are testing changes on their WPCOM sandbox. Using Unison can be more efficient than rsync (or `jetpack rsync`).
+
+  #### Installing Unison
+
+  Please see the respective Unison and usion-fsmonitor repositories for full installation instructions.
```
Comment (daledupreez) [inline docs/development-environment.md]: Typo: `usion` => `unison`
```suggestion
  Please see the respective Unison and unison-fsmonitor repositories for full installation instructions. 
```

File: docs/development-environment.md [hunk 2]
```diff
@@ -192,6 +190,50 @@ There are different types of builds:
 
 	Are pre-commit and pre-push hooks slowing down a major refactor or draft PR? Run `jetpack draft enable` to make them less aggressive (they will still run, but won't block for warnings), and `jetpack draft disable` when you're ready for them again.
 
+* ### Syncing local changes with Unison
```
Comment (daledupreez) [inline docs/development-environment.md]: This item isn't linked in the table of contents - was that intentional?

File: docs/development-environment.md [hunk 3]
```diff
@@ -192,6 +190,50 @@ There are different types of builds:
...
+  
+  In some cases, you may need to test Jetpack (jetpack-mu-wpcom, in particular) changes by syncing your changes to another machine (rather than using Docker). This outlines a strategy for syncing changes in real-time using the [Unison](https://github.com/bcpierce00/unison) file sync tool combines with the [unison-fsmonitor](https://github.com/benesch/unison-fsmonitor).
+
+  This approach may be especially useful for Automatticians who are testing changes on their WPCOM sandbox. Using Unison can be more efficient than rsync (or `jetpack rsync`).
```
Comment (daledupreez) [inline docs/development-environment.md]: I would prefer `WordPress.com` over `WPCOM`. Also, I think it may be worth replacing "more efficient" with "more continuous".

File: docs/development-environment.md [hunk 4]
```diff
@@ -192,6 +190,50 @@ There are different types of builds:
...
+  Please see the respective Unison and usion-fsmonitor repositories for full installation instructions. 
+
+  On OSX, you can use [Homebrew](https://brew.sh/) to quickly install both tools:
+
+  - `brew install unison`
+  - `brew install autozimu/formulas/unison-fsmonitor`
+
+  #### Configuring Unison
+
+  Once Unison is installed, you'll want to create a preferences file. On OSX/Linux, that preferences file would be placed in `~/.unison` and could be called something like `jetpack-plugin-sync.prf`. See the Unison documentation for instructions for other platforms.
```
Comment (daledupreez) [inline docs/development-environment.md:215]: Can we point users to docs via the `unison` command? It looks like there are some decent docs available by running the following commands (with the pipe through `less` to make it easier to scroll/page):
 * `unison -doc tutorial | less`
 * `unison -doc basics | less`
 * `unison -doc running | less`

File: docs/development-environment.md [hunk 5]
```diff
@@ -192,6 +190,50 @@ There are different types of builds:
...
+  
+  Here is a [sample preferences file](unison-sample.prf). Please note that this example preferences file is set to _always_ prefer local changes over remote changes. You'll need to adjust the file if you require a two-way sync instead. See the Unison documentation for full configuration details.
```
Comment (daledupreez) [inline docs/development-environment.md]: This specific section could refer to `unison -doc running | less` as a primer as well.

File: docs/development-environment.md [hunk 6]
```diff
@@ -192,6 +190,50 @@ There are different types of builds:
 
 	Are pre-commit and pre-push hooks slowing down a major refactor or draft PR? Run `jetpack draft enable` to make them less aggressive (they will still run, but won't block for warnings), and `jetpack draft disable` when you're ready for them again.
 
+* ### Syncing local changes with Unison
+  
+  In some cases, you may need to test Jetpack (jetpack-mu-wpcom, in particular) changes by syncing your changes to another machine (rather than using Docker). This outlines a strategy for syncing changes in real-time using the [Unison](https://github.com/bcpierce00/unison) file sync tool combines with the [unison-fsmonitor](https://github.com/benesch/unison-fsmonitor).
```
Comment (daledupreez) [inline docs/development-environment.md]: I think it's _really_ important to note that the server/system you are synching to has `unison` running there as well.

File: docs/development-environment.md [hunk 7]
```diff
@@ -192,6 +190,50 @@ There are different types of builds:
...
+
+  Once your preference file is configured, you can simply run something like the following in a terminal:
+
+  ```
+  unison -ui text -repeat watch jetpack-plugin-sync
+  ```
+
+  Unison will watch for any local changes to the Jetpack files and sync them to your remote host.
+  
+  #### Advanced Configuration
```
Comment (daledupreez) [inline docs/development-environment.md]: This section is completely tied to WordPress.com, and thus feels a bit out of place.

File: docs/development-environment.md [hunk 8]
```diff
@@ -192,6 +190,50 @@ There are different types of builds:
...
+
+  Jetpack currently uses a `sun`/`moon` strategy where the current production files are in one folder and the "staging/test" version are in the other folder.
```
Comment (daledupreez) [inline docs/development-environment.md]: Does the sync work if you simply sync to the already-symlinked `production` directory? That may be simpler.

File: docs/unison-sample.prf [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+ignore = Path .idea
+ignore = Path .nbproject
+ignore = Path .sublime-project
+ignore = Path .sublime-workspace
+ignore = Path .vscode
+ignore = Path changelog
+ignore = Path tests
+
+# Ignore common osx os temp files
+ignore = Name node_modules
```
Comment (daledupreez) [inline docs/unison-sample.prf]: Nit: this isn't an OSX-specific temp file!

File: docs/unison-sample.prf [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+batch = true
+
+# Ignore common IDE files
+ignore = Path .idea
+ignore = Path .nbproject
+ignore = Path .sublime-project
+ignore = Path .sublime-workspace
+ignore = Path .vscode
+ignore = Path changelog
+ignore = Path tests
```
Comment (daledupreez) [inline docs/unison-sample.prf]: These don't seem like IDE-related directories.

File: docs/unison-sample.prf [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+# This can be adjusted to a specific sub-path of the Jetpack repo (like jetpack-mu-wpcom)
+root = /local/path/to/jetpack
+force = /local/path/to/jetpack
+
+# This is the remote host you will be syncing changes to.
+root = ssh://username@remote-host.com//path/to/jetpack/on/remote
+
+batch = true
+
+# Ignore common IDE files
```
Comment (daledupreez) [inline docs/unison-sample.prf]: We should probably prevent mid-build/working/temp files from being synced - this is mostly true for JS(X) and TS(X) files that  get compiled and consolidated during the build process, as that likely occurs in some sort of temporary directory that we wouldn't want to sync to avoid lots of unnecessary changes.

**Review by daledupreez — COMMENTED** [review]
It feels like some elements of the new docs are really WordPress.com-specific and probably belong in a companion/sibling document.

File: docs/unison-sample.prf
Comment (anomiex) [inline docs/unison-sample.prf:1]: Should we move this into `docs/examples/`?

File: docs/development-environment.md [hunk 12]
```diff
@@ -2,31 +2,29 @@
...
-   * [Run a development build](#development-build)
-* [Unit Testing](#unit-testing)
-   * [PHP unit testing](#php-unit-tests)
-   * [JavaScript unit testing](#javascript-unit-tests)
-* [Good code - linting, standards, compatibility, etc.](#good-code---linting-standards-compatibility-etc)
-	* [Coding standards](#coding-standards)
-	* [Linting](#linting)
-* [Standard development & debugging tools](#standard-development--debugging-tools)
+- [Development Environment](#development-environment)
+	- [Table of contents](#table-of-contents)
```
Comment (anomiex) [inline docs/development-environment.md]: Seems kind of pointless to link the TOC from the TOC.

File: docs/development-environment.md [hunk 13]
```diff
@@ -2,31 +2,29 @@
...
+		- [Clone the repository](#clone-the-repository)
+		- [Install development tools](#install-development-tools)
+		- [Check if your environment is ready for Jetpack development](#check-if-your-environment-is-ready-for-jetpack-development)
+- [Development workflow](#development-workflow)
+	- [Building your project](#building-your-project)
+			- [Installing Unison](#installing-unison)
+			- [Configuring Unison](#configuring-unison)
+			- [Running Unison](#running-unison)
+			- [Advanced Configuration](#advanced-configuration)
+	- [Note: You will need to adjust the above command depending on the name(s) of your Unison configuration files.](#note-you-will-need-to-adjust-the-above-command-depending-on-the-names-of-your-unison-configuration-files)
```
Comment (anomiex) [inline docs/development-environment.md]: You removed a bunch of the other sub-section links, but then you're adding way too many for Unison here. What's up with that?

File: docs/development-environment.md (hunk 8 above)
Comment (markbiek) [reply docs/development-environment.md]: Didn't realize that symlink existed, thanks!

File: docs/development-environment.md (hunk 2 above)
Comment (markbiek) [reply docs/development-environment.md]: Nope, I just didn't understand how the ToC links functioned. This is fixed.

File: docs/development-environment.md (hunk 6 above)
Comment (anomiex) [inline docs/development-environment.md]: typo:
```suggestion
  In some cases, you may need to test Jetpack (jetpack-mu-wpcom, in particular) changes by syncing your changes to another machine (rather than using Docker). This outlines a strategy for syncing changes in real-time using the [Unison](https://github.com/bcpierce00/unison) file sync tool combined with the [unison-fsmonitor](https://github.com/benesch/unison-fsmonitor).
```
Also it seems unison-fsmonitor is only for Macs. Might be useful to clarify that, or just omit it here.

File: docs/development-environment.md (hunk 1 above)
Comment (anomiex) [inline docs/development-environment.md]: Linking those installation docs might be a good idea.

File: docs/development-environment.md (hunk 8 above)
Comment (anomiex) [reply docs/development-environment.md]: It does with rsync. No idea about unison.

File: docs/development-environment.md (hunk 7 above)
Comment (anomiex) [reply docs/development-environment.md]: I agree. This seems like something that should have a paYKcK-36O-p2 style reference to a P2 post or field guide page.

File: docs/development-environment.md [hunk 14]
```diff
@@ -190,28 +186,34 @@ There are different types of builds:
...
-* ### Syncing local changes with Unison
+### Syncing local changes with Unison
   
   In some cases, you may need to test Jetpack (jetpack-mu-wpcom, in particular) changes by syncing your changes to another machine (rather than using Docker). This outlines a strategy for syncing changes in real-time using the [Unison](https://github.com/bcpierce00/unison) file sync tool combines with the [unison-fsmonitor](https://github.com/benesch/unison-fsmonitor).
 
-  This approach may be especially useful for Automatticians who are testing changes on their WPCOM sandbox. Using Unison can be more efficient than rsync (or `jetpack rsync`).
+  This approach may be especially useful for Automatticians who are testing changes on their WordPress.com sandbox. Using Unison can be more continuous than rsync (or `jetpack rsync`).
 
-  #### Installing Unison
+  * #### Installing Unison
```
Comment (anomiex) [inline docs/development-environment.md]: Header inside a bullet looks weird when rendered.

File: docs/development-environment.md (hunk 6 above)
Comment (markbiek) [reply docs/development-environment.md]: @daledupreez Is that correct? My understanding was that Unison only ran locally and used an `ssh` connection to sync the files.

File: docs/development-environment.md (hunk 13 above)
Comment (markbiek) [reply docs/development-environment.md]: I wonder if you might have seen that prior to some updates I pushed? This is what it looks like for me now:

CleanShot 2023-09-06 at 11 14 51

Comment (markbiek) [reply docs/development-environment.md]: But I do see what you mean about some of the other sub-section links disappearing. I'll fix those.

Comment (anomiex) [reply docs/development-environment.md]: Yes, it's somewhat better now.

File: docs/development-environment.md (hunk 12 above)
Comment (markbiek) [reply docs/development-environment.md]: That got added by my formatter. I've removed it.

File: docs/development-environment.md (hunk 13 above)
Comment (markbiek) [reply docs/development-environment.md]: I believe the rest is cleaned up now.

File: docs/development-environment.md [hunk 15]
```diff
@@ -2,31 +2,35 @@
...
+			- [PHP](#php)
+			- [Composer](#composer)
+			- [jetpack CLI](#jetpack-cli)
+		- [Check if your environment is ready for Jetpack development](#check-if-your-environment-is-ready-for-jetpack-development)
+- [Development workflow](#development-workflow)
+	- [Building your project](#building-your-project)
+		- [Syncing local changes with Unison](#syncing-local-changes-with-unison)
+			- [Installing Unison](#installing-unison)
+			- [Configuring Unison](#configuring-unison)
+			- [Running Unison](#running-unison)
```
Comment (anomiex) [inline docs/development-environment.md:26]: I don't think we need all these subsections.
```suggestion
```

File: docs/development-environment.md [hunk 16]
```diff
@@ -44,23 +48,23 @@ To get a local WordPress site up and running you need a web server (Apache, Ngin
 
 **Important:** Docker is the only solution that we recommend and can provide support for. The others are listed here as reference if you want to try something different. We won't be able to provide support for them
 
-* ### Docker (Supported Recommended)
+### Docker (Supported Recommended)
```
Comment (anomiex) [inline docs/development-environment.md]: So the bullets here worked ok because the contents of the sections were also part of the bullet. See https://github.com/Automattic/jetpack/blob/c81216cbcf03985541197001d42f3d04fdf2728b/docs/development-environment.md#docker-supported-recommended for example.

Now that you removed the bullets, the spaces at the starts of the sections are making them be displayed as `<pre>` blocks, which we don't want. Look at https://github.com/Automattic/jetpack/blob/f83a5b3f025556c265780ef9d36234036b505ecd/docs/development-environment.md#docker-supported-recommended for example.

If you want to go back and put bullets before your headings in your new section, it would probably work if you do like is done here to have the section contents themselves inside the bullets too.

**Comment by markbiek** [comment]
@anomiex When you have some time, I'd love for you to take another look at this. Hopefully it's ready to merge.

**Comment by anomiex** [comment]
I've been hoping someone else would weigh in, particularly on the question of whether `docs/unison-wordpress-com.md` should be in the monorepo or should be a paYKcK-36O-p2 style reference to a P2 post or field guide page.

File: docs/development-environment.md [hunk 17]
```diff
@@ -192,6 +196,44 @@ There are different types of builds:
...
+
+  Once Unison is installed, you'll want to create a preferences file. On OSX/Linux, that preferences file would be placed in `~/.unison` and could be called something like `jetpack-plugin-sync.prf`. See the Unison documentation for instructions for other platforms.
+
+  The built-in Unison help documentation may be useful:
+
+  - unison -doc tutorial | less
+  - unison -doc basics | less
+  - unison -doc running | less
+  
+  Here is a [sample preferences file](unison-sample.prf). Please note that this example preferences file is set to _always_ prefer local changes over remote changes. You'll need to adjust the file if you require a two-way sync instead. See the Unison documentation (or run `unison -doc running | less`) for full configuration details.
```
Comment (andres-blanco) [inline docs/development-environment.md]: The link for the sample preferences file is broken, it should point into the `examples` directory.

Comment (markbiek) [reply docs/development-environment.md]: Fixed!

File: docs/development-environment.md [hunk 18]
```diff
@@ -192,6 +196,44 @@ There are different types of builds:
...
+  This approach may be especially useful for Automatticians who are testing changes on their WordPress.com sandbox. Using Unison can be more continuous than rsync (or `jetpack rsync`).
+
+  #### Installing Unison
+
+  Please see the respective [Unison](https://github.com/bcpierce00/unison) and [unison-fsmonitor](https://github.com/benesch/unison-fsmonitor) repositories for full installation instructions. 
+
+  On OSX, you can use [Homebrew](https://brew.sh/) to quickly install both tools:
+
+  - `brew install unison`
+  - `brew install autozimu/formulas/unison-fsmonitor`
```
Comment (kraftbj) [inline docs/development-environment.md]:
```suggestion
  - `brew install unison autozimu/formulas/unison-fsmonitor`
```

Can combine them to a single command.

**Review by kraftbj — DISMISSED** [review]
Need to update the PR so this will be dismissed, but the instructions allowed me to functionally use Unison for the first time in my 10 years here :)

**Review by kraftbj — APPROVED** [review]
